File: doc_library/zenodo/supplementary/kspace_substrate_viewer/src/kspace_substrate_viewer.zig
// K-Space Substrate Viewer in Zig + Raylib
// Compile: zig build-exe substrate_viewer.zig -lc -lraylib
// Or use build.zig below

const std = @import("std");
const rl = @cImport({
    @cInclude("raylib.h");
});
const math = std.math;

const ViewState = struct {
    center_x: f32 = 20.0,
    center_y: f32 = 20.0,
    zoom: f32 = 1.0,
};

const WaveSource = struct {
    x: f32,
    y: f32,
    amplitude: f32,
    frequency: f32,
};

const KSpaceSubstrate = struct {
    size: i32,
    data: []f32,
    allocator: std.mem.Allocator,
    texture: rl.Texture2D,
    image_data: []u8,
    needs_update: bool,

    pub fn init(allocator: std.mem.Allocator, size: i32) !KSpaceSubstrate {
        const total = @as(usize, @intCast(size * size));
        const data = try allocator.alloc(f32, total);
        const image_data = try allocator.alloc(u8, total * 4); // RGBA

        var substrate = KSpaceSubstrate{
            .size = size,
            .data = data,
            .allocator = allocator,
            .texture = undefined,
            .image_data = image_data,
            .needs_update = true,
        };

        try substrate.generate();
        substrate.createTexture();

        return substrate;
    }

    pub fn deinit(self: *KSpaceSubstrate) void {
        rl.UnloadTexture(self.texture);
        self.allocator.free(self.data);
        self.allocator.free(self.image_data);
    }

    fn generate(self: *KSpaceSubstrate) !void {
        std.debug.print("Generating {}x{} substrate...\n", .{ self.size, self.size });

        // Wave sources (cymatic interference patterns)
        const sources = [_]WaveSource{
            .{ .x = 10.0, .y = 10.0, .amplitude = 0.8, .frequency = 2.5 },
            .{ .x = 30.0, .y = 10.0, .amplitude = 0.8, .frequency = 2.3 },
            .{ .x = 20.0, .y = 20.0, .amplitude = 0.9, .frequency = 2.7 },
            .{ .x = 10.0, .y = 30.0, .amplitude = 0.7, .frequency = 2.4 },
            .{ .x = 30.0, .y = 30.0, .amplitude = 0.8, .frequency = 2.6 },
        };

        const size_f: f32 = @floatFromInt(self.size);

        var min_val: f32 = math.inf(f32);
        var max_val: f32 = -math.inf(f32);

        // Generate substrate values
        var y: i32 = 0;
        while (y < self.size) : (y += 1) {
            var x: i32 = 0;
            while (x < self.size) : (x += 1) {
                const world_x = (@as(f32, @floatFromInt(x)) / size_f) * 40.0;
                const world_y = (@as(f32, @floatFromInt(y)) / size_f) * 40.0;

                var value: f32 = 0.0;

                // Sum wave sources
                for (sources) |src| {
                    const dx = world_x - src.x;
                    const dy = world_y - src.y;
                    const r = @sqrt(dx * dx + dy * dy);
                    value += src.amplitude * @sin(src.frequency * r) * @exp(-r / 15.0);
                }

                // Add fine structure
                value += 0.3 * @sin(world_x * 0.5) * @cos(world_y * 0.5);
                value += 0.2 * @sin(world_x * 0.8 + world_y * 0.8);

                const idx = @as(usize, @intCast(y * self.size + x));
                self.data[idx] = value;

                if (value < min_val) min_val = value;
                if (value > max_val) max_val = value;
            }
        }

        // Normalize to [-2, 2]
        const range = max_val - min_val;
        for (self.data) |*val| {
            val.* = ((val.* - min_val) / range) * 4.0 - 2.0;
        }

        std.debug.print("Substrate generation complete\n", .{});
    }

    fn createTexture(self: *KSpaceSubstrate) void {
        // Create initial texture (empty, will be updated)
        const img = rl.Image{
            .data = self.image_data.ptr,
            .width = self.size,
            .height = self.size,
            .mipmaps = 1,
            .format = rl.PIXELFORMAT_UNCOMPRESSED_R8G8B8A8,
        };
        self.texture = rl.LoadTextureFromImage(img);
        rl.SetTextureFilter(self.texture, rl.TEXTURE_FILTER_BILINEAR);
    }

    fn colormapDisplacement(value: f32) rl.Color {
        // Normalize from [-2, 2] to [0, 1]
        var normalized = (value + 2.0) / 4.0;
        normalized = math.clamp(normalized, 0.0, 1.0);

        var r: u8 = 0;
        var g: u8 = 0;
        var b: u8 = 0;

        if (normalized < 0.5) {
            // Blue to white
            const t = normalized * 2.0;
            r = @intFromFloat(t * 255.0);
            g = @intFromFloat(t * 255.0);
            b = 255;
        } else {
            // White to red
            const t = (normalized - 0.5) * 2.0;
            r = 255;
            g = @intFromFloat((1.0 - t) * 255.0);
            b = @intFromFloat((1.0 - t) * 255.0);
        }

        return rl.Color{ .r = r, .g = g, .b = b, .a = 255 };
    }

    pub fn updateView(self: *KSpaceSubstrate, view: ViewState, screen_width: i32, screen_height: i32) void {
        // Calculate world space view bounds
        const half_w = (@as(f32, @floatFromInt(screen_width)) / 2.0) / view.zoom;
        const half_h = (@as(f32, @floatFromInt(screen_height)) / 2.0) / view.zoom;

        const world_x_min = view.center_x - half_w;
        const world_x_max = view.center_x + half_w;
        const world_y_min = view.center_y - half_h;
        const world_y_max = view.center_y + half_h;

        // Map to substrate coordinates
        const size_f: f32 = @floatFromInt(self.size);
        const substrate_x_min = @as(i32, @intFromFloat((world_x_min / 40.0) * size_f));
        const substrate_x_max = @as(i32, @intFromFloat((world_x_max / 40.0) * size_f));
        const substrate_y_min = @as(i32, @intFromFloat((world_y_min / 40.0) * size_f));
        const substrate_y_max = @as(i32, @intFromFloat((world_y_max / 40.0) * size_f));

        // Sample and render directly to screen (much faster than texture update)
        var sy: i32 = 0;
        while (sy < screen_height) : (sy += 1) {
            var sx: i32 = 0;
            while (sx < screen_width) : (sx += 1) {
                // Map screen pixel to substrate coordinate
                const t_x = @as(f32, @floatFromInt(sx)) / @as(f32, @floatFromInt(screen_width));
                const t_y = @as(f32, @floatFromInt(sy)) / @as(f32, @floatFromInt(screen_height));

                const sub_x_f = @as(f32, @floatFromInt(substrate_x_min)) + t_x * @as(f32, @floatFromInt(substrate_x_max - substrate_x_min));
                const sub_y_f = @as(f32, @floatFromInt(substrate_y_min)) + t_y * @as(f32, @floatFromInt(substrate_y_max - substrate_y_min));

                const sub_x = math.clamp(@as(i32, @intFromFloat(sub_x_f)), 0, self.size - 1);
                const sub_y = math.clamp(@as(i32, @intFromFloat(sub_y_f)), 0, self.size - 1);

                const idx = @as(usize, @intCast(sub_y * self.size + sub_x));
                const value = self.data[idx];

                const color = colormapDisplacement(value);
                rl.DrawPixel(sx, sy, color);
            }
        }
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Screen settings
    const screen_width: i32 = 1200;
    const screen_height: i32 = 900;

    rl.InitWindow(screen_width, screen_height, "K-Space Substrate Viewer");
    defer rl.CloseWindow();

    rl.SetTargetFPS(60);

    // Generate substrate (2048x2048 for high quality)
    var substrate = try KSpaceSubstrate.init(allocator, 2048);
    defer substrate.deinit();

    // View state
    var view = ViewState{};

    // Movement settings
    const pan_speed: f32 = 0.1;
    const zoom_speed: f32 = 1.05;

    while (!rl.WindowShouldClose()) {
        // Handle input
        const dt = rl.GetFrameTime();
        const move_speed = (pan_speed * 60.0 * dt) / view.zoom;

        // Movement (smooth with delta time)
        if (rl.IsKeyDown(rl.KEY_LEFT) or rl.IsKeyDown(rl.KEY_A)) {
            view.center_x -= move_speed;
        }
        if (rl.IsKeyDown(rl.KEY_RIGHT) or rl.IsKeyDown(rl.KEY_D)) {
            view.center_x += move_speed;
        }
        if (rl.IsKeyDown(rl.KEY_UP) or rl.IsKeyDown(rl.KEY_W)) {
            view.center_y -= move_speed;
        }
        if (rl.IsKeyDown(rl.KEY_DOWN) or rl.IsKeyDown(rl.KEY_S)) {
            view.center_y += move_speed;
        }

        // Zoom
        if (rl.IsKeyPressed(rl.KEY_EQUAL) or rl.IsKeyPressed(rl.KEY_KP_ADD)) {
            view.zoom *= zoom_speed;
        }
        if (rl.IsKeyPressed(rl.KEY_MINUS) or rl.IsKeyPressed(rl.KEY_KP_SUBTRACT)) {
            view.zoom /= zoom_speed;
            view.zoom = @max(0.1, view.zoom);
        }

        // Reset
        if (rl.IsKeyPressed(rl.KEY_R)) {
            view = ViewState{};
        }

        // Clamp view
        view.center_x = math.clamp(view.center_x, 0.0, 40.0);
        view.center_y = math.clamp(view.center_y, 0.0, 40.0);

        // Render
        rl.BeginDrawing();
        defer rl.EndDrawing();

        rl.ClearBackground(rl.BLACK);

        // Draw substrate
        substrate.updateView(view, screen_width, screen_height);

        // Draw UI overlay
        const fps = rl.GetFPS();
        var buffer: [256]u8 = undefined;

        const fps_text = std.fmt.bufPrintZ(&buffer, "FPS: {d}", .{fps}) catch "FPS: --";
        rl.DrawText(fps_text.ptr, 20, 20, 20, rl.WHITE);

        const zoom_text = std.fmt.bufPrintZ(&buffer, "Zoom: {d:.2}x", .{view.zoom}) catch "Zoom: --";
        rl.DrawText(zoom_text.ptr, 20, 45, 20, rl.WHITE);

        const pos_text = std.fmt.bufPrintZ(&buffer, "Pos: ({d:.1}, {d:.1})", .{ view.center_x, view.center_y }) catch "Pos: --";
        rl.DrawText(pos_text.ptr, 20, 70, 20, rl.WHITE);

        rl.DrawText("Arrow/WASD: Move | +/-: Zoom | R: Reset | ESC: Quit", 20, screen_height - 30, 20, rl.WHITE);
    }
}

File: doc_library/zenodo/supplementary/kspace_substrate_viewer.py
"""
K-Space Substrate Viewer
Interactive high-resolution substrate slice visualization
Navigate with arrow keys, zoom with +/-, reset with R
"""

import numpy as np
import pygame
import sys
from dataclasses import dataclass

@dataclass
class ViewState:
    center_x: float = 0.0
    center_y: float = 0.0
    zoom: float = 1.0
    substrate_size: int = 2048  # High resolution substrate
    
class KSpaceSubstrate:
    def __init__(self, size=2048):
        self.size = size
        self.substrate = self._generate_substrate()
        
    def _generate_substrate(self):
        """Generate high-resolution k-space substrate with phase patterns"""
        print(f"Generating {self.size}x{self.size} substrate...")
        
        # Create coordinate grids
        x = np.linspace(0, 40, self.size)
        y = np.linspace(0, 40, self.size)
        X, Y = np.meshgrid(x, y)
        
        # Multiple wave interference patterns (cymatic-like)
        # Create several overlapping wave sources
        substrate = np.zeros((self.size, self.size))
        
        # Wave sources at various positions
        sources = [
            (10, 10, 0.8, 2.5),
            (30, 10, 0.8, 2.3),
            (20, 20, 0.9, 2.7),
            (10, 30, 0.7, 2.4),
            (30, 30, 0.8, 2.6),
        ]
        
        for sx, sy, amplitude, frequency in sources:
            r = np.sqrt((X - sx)**2 + (Y - sy)**2)
            substrate += amplitude * np.sin(frequency * r) * np.exp(-r / 15)
        
        # Add fine structure
        substrate += 0.3 * np.sin(X * 0.5) * np.cos(Y * 0.5)
        substrate += 0.2 * np.sin(X * 0.8 + Y * 0.8)
        
        # Add subtle noise for texture
        substrate += 0.1 * np.random.randn(self.size, self.size)
        
        # Normalize
        substrate = (substrate - substrate.min()) / (substrate.max() - substrate.min())
        substrate = substrate * 4.0 - 2.0  # Scale to [-2, 2] like original
        
        print("Substrate generation complete")
        return substrate
    
    def get_view(self, center_x, center_y, zoom, view_width, view_height):
        """Extract a view from the substrate"""
        # Calculate substrate coordinates
        half_w = (view_width / 2) / zoom
        half_h = (view_height / 2) / zoom
        
        # Map from world coordinates to substrate indices
        substrate_x_min = int((center_x - half_w) * self.size / 40)
        substrate_x_max = int((center_x + half_w) * self.size / 40)
        substrate_y_min = int((center_y - half_h) * self.size / 40)
        substrate_y_max = int((center_y + half_h) * self.size / 40)
        
        # Clamp to substrate bounds
        substrate_x_min = max(0, min(self.size - 1, substrate_x_min))
        substrate_x_max = max(0, min(self.size, substrate_x_max))
        substrate_y_min = max(0, min(self.size - 1, substrate_y_min))
        substrate_y_max = max(0, min(self.size, substrate_y_max))
        
        # Extract region
        if substrate_x_max <= substrate_x_min or substrate_y_max <= substrate_y_min:
            return np.zeros((view_height, view_width))
        
        region = self.substrate[substrate_y_min:substrate_y_max, 
                               substrate_x_min:substrate_x_max]
        
        # Resize to view dimensions using high-quality interpolation
        if region.shape[0] > 0 and region.shape[1] > 0:
            # Use numpy for smooth scaling
            y_scale = view_height / region.shape[0]
            x_scale = view_width / region.shape[1]
            
            # Bilinear interpolation for smooth rendering
            scaled = self._bilinear_scale(region, view_height, view_width)
            return scaled
        
        return np.zeros((view_height, view_width))
    
    def _bilinear_scale(self, img, new_height, new_width):
        """Smooth bilinear scaling"""
        old_height, old_width = img.shape
        
        # Create output array
        scaled = np.zeros((new_height, new_width))
        
        # Calculate scaling ratios
        y_ratio = old_height / new_height
        x_ratio = old_width / new_width
        
        # Sample with bilinear interpolation
        for i in range(new_height):
            for j in range(new_width):
                # Source coordinates (floating point)
                y = i * y_ratio
                x = j * x_ratio
                
                # Integer parts
                y0 = int(y)
                x0 = int(x)
                y1 = min(y0 + 1, old_height - 1)
                x1 = min(x0 + 1, old_width - 1)
                
                # Fractional parts
                dy = y - y0
                dx = x - x0
                
                # Bilinear interpolation
                scaled[i, j] = (
                    img[y0, x0] * (1 - dx) * (1 - dy) +
                    img[y0, x1] * dx * (1 - dy) +
                    img[y1, x0] * (1 - dx) * dy +
                    img[y1, x1] * dx * dy
                )
        
        return scaled

def colormap_displacement(value):
    """Convert displacement value to RGB (blue-white-red colormap)"""
    # Normalize from [-2, 2] to [0, 1]
    normalized = (value + 2.0) / 4.0
    normalized = np.clip(normalized, 0, 1)
    
    if normalized < 0.5:
        # Blue to white
        t = normalized * 2
        r = int(t * 255)
        g = int(t * 255)
        b = 255
    else:
        # White to red
        t = (normalized - 0.5) * 2
        r = 255
        g = int((1 - t) * 255)
        b = int((1 - t) * 255)
    
    return (r, g, b)

def main():
    # Initialize Pygame
    pygame.init()
    
    # Screen settings
    SCREEN_WIDTH = 1200
    SCREEN_HEIGHT = 900
    screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
    pygame.display.set_caption("K-Space Substrate Viewer")
    
    # View state
    view = ViewState(center_x=20.0, center_y=20.0, zoom=1.0)
    
    # Generate substrate
    substrate = KSpaceSubstrate(size=2048)
    
    # Movement settings
    pan_speed = 0.5  # World units per keypress
    zoom_speed = 1.2
    
    # FPS
    clock = pygame.time.Clock()
    font = pygame.font.Font(None, 36)
    
    running = True
    while running:
        # Handle events
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    running = False
                elif event.key == pygame.K_r:
                    # Reset view
                    view.center_x = 20.0
                    view.center_y = 20.0
                    view.zoom = 1.0
                elif event.key == pygame.K_EQUALS or event.key == pygame.K_PLUS:
                    view.zoom *= zoom_speed
                elif event.key == pygame.K_MINUS:
                    view.zoom /= zoom_speed
                    view.zoom = max(0.1, view.zoom)
        
        # Continuous key handling for smooth movement
        keys = pygame.key.get_pressed()
        move_speed = pan_speed / view.zoom
        
        if keys[pygame.K_LEFT] or keys[pygame.K_a]:
            view.center_x -= move_speed
        if keys[pygame.K_RIGHT] or keys[pygame.K_d]:
            view.center_x += move_speed
        if keys[pygame.K_UP] or keys[pygame.K_w]:
            view.center_y -= move_speed
        if keys[pygame.K_DOWN] or keys[pygame.K_s]:
            view.center_y += move_speed
        
        # Clamp view to substrate bounds
        view.center_x = np.clip(view.center_x, 0, 40)
        view.center_y = np.clip(view.center_y, 0, 40)
        
        # Get current view
        view_data = substrate.get_view(
            view.center_x, view.center_y, view.zoom,
            SCREEN_WIDTH, SCREEN_HEIGHT
        )
        
        # Convert to RGB
        rgb_array = np.zeros((SCREEN_HEIGHT, SCREEN_WIDTH, 3), dtype=np.uint8)
        for i in range(SCREEN_HEIGHT):
            for j in range(SCREEN_WIDTH):
                rgb_array[i, j] = colormap_displacement(view_data[i, j])
        
        # Blit to screen
        surf = pygame.surfarray.make_surface(rgb_array.swapaxes(0, 1))
        screen.blit(surf, (0, 0))
        
        # Draw UI overlay
        fps = clock.get_fps()
        fps_text = font.render(f"FPS: {fps:.1f}", True, (255, 255, 255))
        zoom_text = font.render(f"Zoom: {view.zoom:.2f}x", True, (255, 255, 255))
        pos_text = font.render(f"Pos: ({view.center_x:.1f}, {view.center_y:.1f})", 
                              True, (255, 255, 255))
        
        # Draw semi-transparent background for text
        info_rect = pygame.Rect(10, 10, 300, 100)
        s = pygame.Surface((300, 100))
        s.set_alpha(128)
        s.fill((0, 0, 0))
        screen.blit(s, (10, 10))
        
        screen.blit(fps_text, (20, 20))
        screen.blit(zoom_text, (20, 50))
        screen.blit(pos_text, (20, 80))
        
        # Draw controls hint
        hint_text = font.render("Arrow keys: Move | +/-: Zoom | R: Reset | ESC: Quit", 
                               True, (255, 255, 255))
        hint_rect = pygame.Rect(10, SCREEN_HEIGHT - 50, SCREEN_WIDTH - 20, 40)
        s2 = pygame.Surface((SCREEN_WIDTH - 20, 40))
        s2.set_alpha(128)
        s2.fill((0, 0, 0))
        screen.blit(s2, (10, SCREEN_HEIGHT - 50))
        screen.blit(hint_text, (20, SCREEN_HEIGHT - 40))
        
        pygame.display.flip()
        clock.tick(60)  # Target 60 FPS
    
    pygame.quit()
    sys.exit()

if __name__ == "__main__":
    main()

